app/controllers/concerns/checkout_callbacks.rb

```diff
@@ -47,7 +50,25 @@ def load_saved_credit_cards
   end
 
   def load_shipping_methods
-    @shipping_methods = available_shipping_methods.sort { |a, b| a.name.casecmp(b.name) }
+    @shipping_methods = sorted_available_shipping_methods
+  end
+
+  def load_allowed_shipping_methods
+    @allowed_shipping_methods = sorted_available_shipping_methods.filter(
+      &method(:supports_all_products_shipping_categories?)
+    )
+  end
+
+  def sorted_available_shipping_methods
+    available_shipping_methods.sort { |a, b| a.name.casecmp(b.name) }
+  end
+
+  def supports_all_products_shipping_categories?(shipping_method)
+    (products_shipping_categories - shipping_method.shipping_categories.pluck(:id)).empty?
```

Comment:
Clever. It would be even clearer to use a set:
```suggestion
    products_shipping_categories.subset?(shipping_method.shipping_categories.pluck(:id))
```

Reply:
```ruby
3.0.3 :065 > [].subset
(irb):65:in `<main>': undefined method `subset' for []:Array (NoMethodError)
```